README.md
# wg-access-server

wg-access-server is a single binary file that contains a WireGuard
VPN server and a web user interface for device management. We support user authentication,
_1-click_ device enrollment that works with macOS, Linux, Windows, iOS/iPadOS and Android
including QR codes. Furthermore, you can choose from different network isolation modes for a
better control over connected devices. Generally speaking you can customize the project
to your use-case with relative ease.

This project aims to provide a simple VPN solution for developers,
homelab enthusiasts, and anyone else who is adventurous.

**This is a fork of the original work of place1, maintained by [Freifunk Munich](https://ffmuc.net/).
Since the upstream is currently unmaintained, we try to add new features and keep the project up to date and in a working state.**

This fork supports IPv6. The VPN can run in dual-stack, IPv6-only or IPv4-only mode.
NAT can be disabled separately for IPv4 and IPv6.

**Contributions are always welcome so that we can offer new bug fixes, features and improvements to the users of this project**.

## Features

- Pluggable authentication using OpenID Connect
- Authentication using GitLab
- IPv6 support in tunnel
- Caching DNS proxy (stub resolver)
- WireGuard client configuration QR codes
- PostgreSQL, MySQL or SQLite3 storage backend

## Documentation

[See our documentation website](https://www.freie-netze.org/wg-access-server/)

Quick Links:

- [Configuration Overview](https://www.freie-netze.org/wg-access-server/2-configuration/)
- [Deploy With Docker](https://www.freie-netze.org/wg-access-server/deployment/1-docker/)
- [Deploy With Docker-Compose](https://www.freie-netze.org/wg-access-server/deployment/2-docker-compose/)
- [Deploy With Helm](https://www.freie-netze.org/wg-access-server/deployment/3-kubernetes/)

## Running with Docker

Here is a quick command to start the wg-access-server for the first time and try it out.

```bash
export WG_ADMIN_PASSWORD="example"
export WG_WIREGUARD_PRIVATE_KEY="$(wg genkey)"

docker run \
  -it \
  --rm \
  --cap-add NET_ADMIN \
  --device /dev/net/tun:/dev/net/tun \
  --sysctl net.ipv6.conf.all.disable_ipv6=0 \
  --sysctl net.ipv6.conf.all.forwarding=1 \
  -v wg-access-server-data:/data \
  -v /lib/modules:/lib/modules \
  -e "WG_ADMIN_PASSWORD=$WG_ADMIN_PASSWORD" \
  -e "WG_WIREGUARD_PRIVATE_KEY=$WG_WIREGUARD_PRIVATE_KEY" \
  -p 8000:8000/tcp \
  -p 51820:51820/udp \
  ghcr.io/freifunkmuc/wg-access-server:latest
```

If the wg-access-server is accessible via LAN or a network you are in, you can directly connect your phone to the VPN. You have to call the webfrontent of the project for this. Normally, this is done via the IP address of the device or server on which the wg-access-server is running followed by the standard port 8000, via which the web interface can be reached. For most deployments something like this should work: http://192.168.0.XX:8000

If the project is running locally on the computer, you can easily connect to the web interface by connecting to http://localhost:8000 in the browser.

## Running on Kubernetes via Helm

wg-access-server ships a Helm chart to make it easy to get started on
Kubernetes.

Here's a quick start, but you can read more at the [Helm Chart Deployment Docs](https://freifunkMUC.github.io/wg-access-server/deployment/3-kubernetes/)

```bash
# deploy
helm install my-release --repo https://freifunkMUC.github.io/wg-access-server wg-access-server

# cleanup
helm delete my-release
```

## Running with Docker-Compose

Download the the docker-compose.yml file from the repo and run the following command.

```bash
export WG_ADMIN_PASSWORD="example"
export WG_WIREGUARD_PRIVATE_KEY="$(wg genkey)"

docker-compose up
```

You can connect to the web server on the local machine browser at http://localhost:8000

If you open your browser to your machine's LAN IP address you'll be able
to connect your phone using the UI and QR code!

## Screenshots

![Devices](https://github.com/freifunkMUC/wg-access-server/raw/master/screenshots/devices.png)

![Connect iOS](https://github.com/freifunkMUC/wg-access-server/raw/master/screenshots/connect-mobile.png)

![Connect MacOS](https://github.com/freifunkMUC/wg-access-server/raw/master/screenshots/connect-desktop.png)

![Sign In](https://github.com/freifunkMUC/wg-access-server/raw/master/screenshots/signin.png)

## Changelog

See the [CHANGELOG.md](https://github.com/freifunkMUC/wg-access-server/blob/master/CHANGELOG.md) file

## Development

The software consists of a Golang server and a React app.

If you want to make changes to the project locally, you can do so relatively easily with the following steps.

1. Run `cd website && npm install && npm start` to get the frontend running on `:3000`.
2. Run `sudo go run ./main.go` to get the server running on `:8000`.

Here are some notes on development configuration:

- sudo is required because the server uses iptables/ip to configure the VPN network
- access to the website is on `:3000` and API requests are redirected to `:8000` thanks to webpack
- in-memory storage and generated WireGuard keys are used

gRPC code generation:

The client communicates with the server via gRPC web. You can edit the API specification in `./proto/*.proto`.

After changing a service or message definition, you must regenerate the server and client code using: `./codegen.sh`.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Latest]

- Added initial support for IPv6
- Allow disabling of IPv4 or IPv6
- Added NAT66
- Added dependabot for easier updates of dependencies
- Added gh-actions for the build process of the image itself
- Utilizing GitHub Cache for faster image building
- Added support for arm64, arm/v7
- Added OpenContainers labels to the build process (gh-actions)
- Added proper go linter
- Fixed DNS proxy upstream lookup loop
- Updated the documentation to reflect the changes regarding IPv6

## [v0.4.6]

- The docker compose file now works without a config file (it's optional)

## [v0.4.5]

- Fixed a bug that caused devices to not correctly sync when using mysql or sqlite3

## [v0.4.4]

### Changed

- The default AllowedIPs setting was changed from "0.0.0.0/1, 128.0.0.0/1" to "0.0.0.0/0".

## [v0.4.3]

### Changed

- The device list on the website now updates a little less frequently.
- The device list now always shows the "last seen" field to hopefully
  better reflect what the "connected" status means.
- The metadata scraping loop has been updated to be more efficient when
  there are many disconnected peers compared to connected peers.
- The metadata scraping algorithm is now more friendly for HA deployments.

## [v0.4.2]

### Bug Fixes

- The vpn Allowed IPs setting is now correctly enforced.

## [v0.4.1]

### Bug Fixes

- Fixed a bug that caused devices to get disconnected intermittently
- The helm template now respects the "replicas" value

## [v0.4.0]

### Added

- High availability (HA) is now supported when using the `postgresql://` storage backend.
  You can now deploy multiple replicas of wg-access-server pointing to the same Postgres DB.
- The wireguard service can now be disabled via the config file. Helpful for developing
  on Mac and Windows.

### Removed

- The `file://` storage backend was deprecated in v0.3.0 and has now been removed.
  See the v0.3.0 changelog entry for more information about migrating your data.

## [v0.3.0]

### Added

- arm64 and arm/v7 docker image support + github actions thanks to [@timtorChen](https://github.com/Place1/wg-access-server/pull/73)

### Changed

- the wireguard private key is now required when the storage backend is persistent (i.e. not `memory://`)
- configuration flags, environment variables and file properties have been refactored for consistency
  * all configuration file properties (excluding auth providers) can now be set via flags and environment variables
  * all environment variables are prefixed with `WG_` to avoid collisions in hosted environments like Kubernetes
  * all flags & environment variables are named consistently
  * **breaking:** no functionality has been removed but you'll need to update any flags/envvars that you're using

### Deprecations

- deprecated support for having no admin account
  * a config error will be thrown in v0.4.0 if an admin account is not configured
  * see the README.md for examples on setting the admin account
- deprecated `file://` storage in favour of `sqlite3://`
  * will be removed in v0.4.0
  * there is now a storage `migrate` command that you can use to move your data to a different storage backend
  * see the docs for migrating your data: https://place1.github.io/wg-access-server/3-storage/#example-file-to-sqlite3

## [0.2.5]

### Added

- Admin users can now delete devices from the "all devices" page (issue [#57](https://github.com/Place1/wg-access-server/issues/57))

### Bug Fixes

- Fixes website routing to solve 404s (issue [#56](https://github.com/Place1/wg-access-server/issues/56))

## [0.2.4]

### Bug Fixes

- Improved config validation and error reporting (issue [#58](https://github.com/Place1/wg-access-server/issues/58) [#61](https://github.com/Place1/wg-access-server/issues/61))

## [0.2.3]

### Added

- Helm chart now supports configuring a LoadBalancer service for the web ui ([@nqngo](https://github.com/Place1/wg-access-server/pull/60))

## [0.2.2]

### Changed

- Changed the default "AllowedIPs" to `0.0.0.0/0`

## [0.2.1]

### Changed

- The "is connected" now shows devices as connected if they've been active within the last 3 minutes
- Improved handling of oidc/gitlab authentication with domain verification when a user hasn't set their email

## [0.2.0]

### Added

- New SQL storage backend supporting SQLite, MySQL and PostgreSQL ([@halkeye](https://github.com/Place1/wg-access-server/pull/37))
- Support for mapping claims from an OIDC auth backend to wg-access-server claims using a simple rule
  syntax ([@halkeye](https://github.com/Place1/wg-access-server/pull/39)). You can use this feature
  to decide which user has the 'admin' claim based on your own OIDC claims.
- The VPN DNS proxy feature can now be disabled using config: `dns.enabled = false`
  - When disabled the `DNS` wireguard config value will be omitted from client wg config files
  - When disabled the DNSasd proxy will not be started server-side (i.e. port 53 won't be used)
- Config options to change the web, wireguard and dns ports.
- Better instructions for connecting a linux device ([@nfg](https://github.com/Place1/wg-access-server/pull/38))
- More helm chart flexibility ([@halkeye](https://github.com/Place1/wg-access-server/pull/33))

### Changes

- The admin UI will now show the device owner's name or email if available.
- The admin UI will now show the auth provider for a given device if more than 1 auth provider is in use.
- Bug fix: upstream dns now correctly configured using resolvconf if not set in config file, flag or envvar.

### Removed

- dns port configuration was removed because wireguard client's only support port 53 for dns

### How to upgrade

- If you've been using the `storage.directory="/some/path"` config value then

cmd/serve/main.go
package serve

import (
	"fmt"
	"io/ioutil"
	"net"
	"net/http"
	"strings"

	"github.com/docker/libnetwork/resolvconf"
	"github.com/docker/libnetwork/types"
	"github.com/place1/wg-access-server/internal/services"
	"github.com/place1/wg-access-server/internal/storage"
	"github.com/place1/wg-access-server/pkg/authnz"
	"github.com/place1/wg-access-server/pkg/authnz/authconfig"
	"github.com/place1/wg-access-server/pkg/authnz/authsession"
	"github.com/vishvananda/netlink"
	"golang.org/x/crypto/bcrypt"
	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
	"gopkg.in/alecthomas/kingpin.v2"
	"gopkg.in/yaml.v2"

	"github.com/gorilla/mux"
	"github.com/place1/wg-embed/pkg/wgembed"

	"github.com/pkg/errors"
	"github.com/place1/wg-access-server/internal/config"
	"github.com/place1/wg-access-server/internal/devices"
	"github.com/place1/wg-access-server/internal/dnsproxy"
	"github.com/place1/wg-access-server/internal/network"
	"github.com/sirupsen/logrus"
)

func Register(app *kingpin.Application) *servecmd {
	cmd := &servecmd{}
	cli := app.Command(cmd.Name(), "Run the server")
	cli.Flag("config", "Path to a wg-access-server config file").Envar("WG_CONFIG").StringVar(&cmd.ConfigFilePath)
	cli.Flag("admin-username", "Admin username (defaults to admin)").Envar("WG_ADMIN_USERNAME").Default("admin").StringVar(&cmd.AppConfig.AdminUsername)
	cli.Flag("admin-password", "Admin password (provide plaintext, stored in-memory only)").Envar("WG_ADMIN_PASSWORD").StringVar(&cmd.AppConfig.AdminPassword)
	cli.Flag("port", "The port that the web ui server will listen on").Envar("WG_PORT").Default("8000").IntVar(&cmd.AppConfig.Port)
	cli.Flag("external-host", "The external origin of the server (e.g. https://mydomain.com)").Envar("WG_EXTERNAL_HOST").StringVar(&cmd.AppConfig.ExternalHost)
	cli.Flag("storage", "The storage backend connection string").Envar("WG_STORAGE").Default("memory://").StringVar(&cmd.AppConfig.Storage)
	cli.Flag("disable-metadata", "Disable metadata collection (i.e. metrics)").Envar("WG_DISABLE_METADATA").Default("false").BoolVar(&cmd.AppConfig.DisableMetadata)
	cli.Flag("wireguard-enabled", "Enable or disable the embedded wireguard server (useful for development)").Envar("WG_WIREGUARD_ENABLED").Default("true").BoolVar(&cmd.AppConfig.WireGuard.Enabled)
	cli.Flag("wireguard-interface", "Set the wireguard interface name").Default("wg0").Envar("WG_WIREGUARD_INTERFACE").StringVar(&cmd.AppConfig.WireGuard.Interface)
	cli.Flag("wireguard-private-key", "Wireguard private key").Envar("WG_WIREGUARD_PRIVATE_KEY").StringVar(&cmd.AppConfig.WireGuard.PrivateKey)
	cli.Flag("wireguard-port", "The port that the Wireguard server will listen on").Envar("WG_WIREGUARD_PORT").Default("51820").IntVar(&cmd.AppConfig.WireGuard.Port)
	cli.Flag("vpn-cidr", "The network CIDR for the VPN").Envar("WG_VPN_CIDR").Default("10.44.0.0/24").StringVar(&cmd.AppConfig.VPN.CIDR)
	cli.Flag("vpn-cidrv6", "The IPv6 network CIDR for the VPN").Envar("WG_VPN_CIDRV6").Default("fd48:4c4:7aa9::/64").StringVar(&cmd.AppConfig.VPN.CIDRv6)
	cli.Flag("vpn-nat44-enabled", "Enable or disable NAT of IPv6 traffic leaving through the gateway").Envar("WG_IPV4_NAT_ENABLED").Default("true").BoolVar(&cmd.AppConfig.VPN.NAT44)
	cli.Flag("vpn-nat66-enabled", "Enable or disable NAT of IPv6 traffic leaving through the gateway").Envar("WG_IPV6_NAT_ENABLED").Default("true").BoolVar(&cmd.AppConfig.VPN.NAT66)
	cli.Flag("vpn-gateway-interface", "The gateway network interface (i.e. eth0)").Envar("WG_VPN_GATEWAY_INTERFACE").Default(detectDefaultInterface()).StringVar(&cmd.AppConfig.VPN.GatewayInterface)
	cli.Flag("vpn-allowed-ips", "A list of networks that VPN clients will be allowed to connect to via the VPN").Envar("WG_VPN_ALLOWED_IPS").Default("0.0.0.0/0", "::/0").StringsVar(&cmd.AppConfig.VPN.AllowedIPs)
	cli.Flag("dns-enabled", "Enable or disable the embedded dns proxy server (useful for development)").Envar("WG_DNS_ENABLED").Default("true").BoolVar(&cmd.AppConfig.DNS.Enabled)
	cli.Flag("dns-upstream", "An upstream DNS server to proxy DNS traffic to. Defaults to resolveconf with Cloudflare DNS as fallback").Envar("WG_DNS_UPSTREAM").StringsVar(&cmd.AppConfig.DNS.Upstream)
	return cmd
}

type servecmd struct {
	ConfigFilePath string
	AppConfig      config.AppConfig
}

func (cmd *servecmd) Name() string {
	return "serve"
}

func (cmd *servecmd) Run() {
	conf := cmd.ReadConfig()

	if conf.VPN.CIDR == "0" {
		conf.VPN.CIDR = ""
	}
	if conf.VPN.CIDRv6 == "0" {
		conf.VPN.CIDRv6 = ""
	}

	// Get the server's IP addresses within the VPN
	var vpnip, vpnipv6 *net.IPNet
	var err error
	vpnip, vpnipv6, err = network.ServerVPNIPs(conf.VPN.CIDR, conf.VPN.CIDRv6)
	if err != nil {
		logrus.Fatal(err)
	}
	if vpnip == nil && vpnipv6 == nil {
		logrus.Fatal("need at least one of VPN.CIDR or VPN.CIDRv6 set")
	}

	// Allow traffic to wg-access-server's peer endpoint.
	// This is important because clients will send traffic
	// to the embedded DNS proxy using the VPN IP
	vpnipstrings := make([]string, 0, 2)
	if vpnip != nil {
		conf.VPN.AllowedIPs = append(conf.VPN.AllowedIPs, fmt.Sprintf("%s/32", vpnip.IP.String()))
		vpnipstrings = append(vpnipstrings, vpnip.String())
	}
	if vpnipv6 != nil {
		conf.VPN.AllowedIPs = append(conf.VPN.AllowedIPs, fmt.Sprintf("%s/128", vpnipv6.IP.String()))
		vpnipstrings = append(vpnipstrings, vpnipv6.String())
	}

	// WireGuard Server
	wg := wgembed.NewNoOpInterface()
	if conf.WireGuard.Enabled {
		wgimpl, err := wgembed.New(conf.WireGuard.Interface)
		if err != nil {
			logrus.Fatal(errors.Wrap(err, "failed to create wireguard interface"))
		}
		defer wgimpl.Close()
		wg = wgimpl

		logrus.Infof("starting wireguard server on :%d", conf.WireGuard.Port)

		wgconfig := &wgembed.ConfigFile{
			Interface: wgembed.IfaceConfig{
				PrivateKey: conf.WireGuard.PrivateKey,
				Address:    vpnipstrings,
				ListenPort: &conf.WireGuard.Port,
			},
		}

		if err := wg.LoadConfig(wgconfig); err != nil {
			logrus.Fatal(errors.Wrap(err, "failed to load wireguard config"))
		}

		logrus.Infof("wireguard VPN network is %s", network.StringJoinIPNets(vpnip, vpnipv6))

		if err := network.ConfigureForwarding(conf.VPN.GatewayInterface, conf.VPN.CIDR, conf.VPN.CIDRv6, conf.VPN.NAT44, conf.VPN.NAT66, conf.VPN.AllowedIPs); err != nil {
			logrus.Fatal(err)
		}
	}

	// DNS Server
	if conf.DNS.Enabled {
		if conf.DNS.Upstream == nil {
			conf.DNS.Upstream = detectDNSUpstream(conf.VPN.CIDR != "", conf.VPN.CIDRv6 != "")
		}
		dns, err := dnsproxy.New(dnsproxy.DNSServerOpts{
			Upstream: conf.DNS.Upstream,
		})
		if err != nil {
			logrus.Fatal(errors.Wrap(err, "failed to start dns server"))
		}
		defer dns.Close()
	}

	// Storage
	storageBackend, err := storage.NewStorage(conf.Storage)
	if err != nil {
		logrus.Fatal(errors.Wrap(err, "failed to create storage backend"))
	}
	if err := storageBackend.Open(); err != nil {
		logrus.Fatal(errors.Wrap(err, "failed to connect/open storage backend"))
	}
	defer storageBackend.Close()

	// Services
	deviceManager := devices.New(wg, storageBackend, conf.VPN.CIDR, conf.VPN.CIDRv6)
	if err := deviceManager.StartSync(conf.DisableMetadata); err != nil {
		logrus.Fatal(errors.Wrap(err, "failed to sync"))
	}

	router := mux.NewRouter()
	router.Use(services.TracesMiddleware)
	router.Use(services.RecoveryMiddleware)

	// Health check endpoint
	router.PathPrefix("/health").Handler(services.HealthEndpoint())

	// Authentication middleware
	if conf.Auth.IsEnabled() {
		middleware, err := authnz.NewMiddleware(conf.Auth, claimsMiddleware(conf))
		if err != nil {
			logrus.Fatal(errors.Wrap(err, "failed to set up authnz middleware"))
		}
		router.Use(middleware)
	} else {
		logrus.Warn("[DEPRECATION NOTICE] using wg-access-server without an admin user is deprecated and will be removed in an upcoming minor release.")
		router.Use(func(next http.Handler) http.Handler {
			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				next.ServeHTTP(w, r.WithContext(authsession.SetIdentityCtx(r.Context(), &authsession.AuthSession{
					Identity: &authsession.Identity{
						Subject: "",
					},
				})))
			})
		})
	}

	// Subrouter for our site (web + api)
	site := router.PathPrefix("/").Subrouter()
	site.Use(authnz.RequireAuthentication)

	// Grpc api
	site.PathPrefix("/api").Handler(services.ApiRouter(&services.ApiServices{
		Config:        conf,
		DeviceManager: deviceManager,
		Wg:            wg,
	}))

	// Static website
	site.PathPrefix("/").Handler(services.WebsiteRouter())

	publicRouter := router

	// Listen
	address := fmt.Sprintf(":%d", conf.Port)
	srv := &http.Server{
		Addr:    address,
		Handler: publicRouter,
	}

	// Start Web server
	logrus.Infof("web ui listening on %v", address)
	if err := srv.ListenAndServe(); err != nil {
		logrus.Fatal(errors.Wrap(err, "unable to start http server"))
	}
}

func (cmd *servecmd) ReadConfig() *config.AppConfig {
	if cmd.ConfigFilePath != "" {
		if b, err := ioutil.ReadFile(cmd.ConfigFilePath); err == nil {
			if err := yaml.Unmarshal(b, &cmd.AppConfig); err != nil {
				logrus.Fatal(errors.Wrap(err, "failed to bind configuration file"))
			}
		}
	}

	if cmd.AppConfig.LogLevel != "" {
		if level, err := logrus.ParseLevel(cmd.AppConfig.LogLevel); err == nil {
			logrus.SetLevel(level)
		}
	}

	if cmd.AppConfig.AdminPassword == "" {
		logrus.Fatal("missing admin password: please set via environment variable, flag or config file")
	}

	if cmd.AppConfig.DisableMetadata {
		logrus.Info("Metadata collection has been disabled. No metrics or device connectivity information will be recorded or shown")
	}

	// set a basic auth entry for the admin user
	if cmd.AppConfig.Auth.Basic == nil {
		cmd.AppConfig.Auth.Basic = &authconfig.BasicAuthConfig{}
	}
	pw, err := bcrypt.GenerateFromPassword([]byte(cmd.AppConfig.AdminPassword), bcrypt.DefaultCost)
	if err != nil {
		logrus.Fatal(errors.Wrap(err, "failed to generate a bcrypt hash for the provided admin password"))
	}
	cmd.AppConfig.Auth.Basic.Users = append(cmd.AppConfig.Auth.Basic.Users, fmt.Sprintf("%s:%s", cmd.AppConfig.AdminUsername, string(pw)))

	// we'll generate a private key when using memory://
	// storage only.
	if cmd.AppConfig.WireGuard.PrivateKey == "" {
		if !strings.HasPrefix(cmd.AppConfig.Storage, "memory://") {
			logrus.Fatal(missingPrivateKey)
		}
		key, err := wgtypes.GeneratePrivateKey()
		if err != nil {
			logrus.Fatal(errors.Wrap(err, "failed to generate a server private key"))
		}
		cmd.AppConfig.WireGuard.PrivateKey = key.String()
	}

	return &cmd.AppConfig
}

func claimsMiddleware(conf *config.AppConfig) authsession.ClaimsMiddleware {
	return func(user *authsession.Identity) error {
		if user.Subject == conf.AdminUsername {
			user.Claims.Add("admin", "true")
		}
		return nil
	}
}

func detectDNSUpstream(ipv4Enabled, ipv6Enabled bool) []string {
	upstream := []string{}
	if r, err := resolvconf.Get(); err == nil {
		upstream = resolvconf.GetNameservers(r.Content, types.IP)
	}
	if len(upstream) == 0 {
		logrus.Warn("failed to get nameservers from /etc/resolv.conf defaulting to Cloudflare DNS instead")
		// If there's no default route for IPv6, lookup fails immediately without delay and we retry using IPv4
		if ipv6Enabled {
			upstream = append(upstream, "2606:4700:4700::1111")
		}
		if ipv4Enabled {
			upstream = append(upstream, "1.1.1.1")
		}
	}
	return upstream
}

func detectDefaultInterface() string {
	links, err := netlink.LinkList()
	if err != nil {
		logrus.Warn(errors.Wrap(err, "failed to list network interfaces"))
		return ""
	}
	for _, link := range links {
		// First try IPv4, then IPv6, hope both have the same default interface
		for family := range []int{4, 6} {
			routes, err := netlink.RouteList(link, family)
			if err != nil {
				logrus.Warn(errors.Wrapf(err, "failed to list routes for interface %s", link.Attrs().Name))
				return ""
			}
			for _, route := range routes {
				if route.Dst == nil {
					return link.Attrs().Name
				}
			}
		}
	}
	logrus.Warn(errors.New("could not determine the default network interface name"))
	return ""
}

var missingPrivateKey = `missing wireguard private key:

    create a key:

        $ wg genkey

    configure via environment variable:

        $ export WIREGUARD_PRIVATE_KEY="<private-key>"

    or configure via flag:

        $ wg-access-server serve --wireguard-private-key="<private-key>"

    or configure via file:

      wireguard:
        privateKey: "<private-key>"

`
